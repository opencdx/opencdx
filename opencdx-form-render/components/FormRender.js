import React from 'react';
import { useState } from 'react';
import { StyleSheet } from 'react-native';
import { Button, ButtonText, Input, InputField, Radio, RadioGroup, RadioIndicator, RadioIcon, CircleIcon, HStack, RadioLabel, Select, SelectTrigger, SelectItem, SelectIcon, SelectContent, SelectPortal, SelectDragIndicatorWrapper, SelectBackdrop, SelectInput, Icon, ChevronDownIcon, SelectDragIndicator, FormControl, FormControlLabel, FormControlLabelText, Heading } from '@gluestack-ui/themed';
import formData from './alpha.json';

const FormRender = React.forwardRef((props, ref) => {
    const [cuff, setCuff] = useState();
    const [urinated, setUrinated] = useState();
    const [sitting, setSitting] = useState();

    return(
        <FormControl minWidth="$80">
            <Heading>{formData?.title}</Heading>
            { formData?.item?.map((question, index) => {
                return(
                    <>
                    <FormControlLabel>
                        <FormControlLabelText>{question.text}</FormControlLabelText>
                    </FormControlLabel>

                    { question.type === "integer" &&
                        <Input
                            style={styles.margin}
                            variant="outline"
                            size="md"
                            isDisabled={false}
                            isInvalid={false}
                            isReadOnly={false}
                        >
                            <InputField placeholder="Enter Text here" keyboardType="numeric"/>
                        </Input>
                    }

                    { question.type === "string" &&
                        <Input
                            style={styles.margin}
                            variant="outline"
                            size="md"
                            isDisabled={false}
                            isInvalid={false}
                            isReadOnly={false}
                        >
                            <InputField placeholder="Enter Text here" />
                        </Input>
                    }

                    { question.type === "text" &&
                        <Input
                            style={styles.margin}
                            variant="outline"
                            size="md"
                            isDisabled={false}
                            isInvalid={false}
                            isReadOnly={false}
                        >
                            <InputField placeholder="Enter Text here" />
                        </Input>
                    }

                    { question.type === "boolean" &&
                        <RadioGroup value={cuff} onChange={setCuff} style={styles.margin}>
                            <HStack space="2xl">
                                <Radio value="yes">
                                <RadioIndicator mr="$2">
                                    <RadioIcon as={CircleIcon} />
                                </RadioIndicator>
                                <RadioLabel>Yes</RadioLabel>
                                </Radio>
                                <Radio value="no">
                                <RadioIndicator mr="$2">
                                    <RadioIcon as={CircleIcon} />
                                </RadioIndicator>
                                <RadioLabel>No</RadioLabel>
                                </Radio>
                                <Radio value="notAnswered">
                                <RadioIndicator mr="$2">
                                    <RadioIcon as={CircleIcon} />
                                </RadioIndicator>
                                <RadioLabel>Not Answered</RadioLabel>
                                </Radio>
                            </HStack>
                        </RadioGroup>
                    }

                    { question.type === "choice" &&
                        <Select style={styles.margin}>
                            <SelectTrigger variant="outline" size="md">
                                <SelectInput placeholder="Select option" />
                                <SelectIcon mr="$3">
                                <Icon as={ChevronDownIcon} />
                                </SelectIcon>
                            </SelectTrigger>
                            <SelectPortal>
                                <SelectBackdrop />
                                <SelectContent>
                                <SelectDragIndicatorWrapper>
                                    <SelectDragIndicator />
                                </SelectDragIndicatorWrapper>
                                {question.answerOption.map((option) => {
                                    return(
                                        <SelectItem label={option.valueCoding.display} value={option.valueCoding.display} />
                                    )
                                })}
                                </SelectContent>
                            </SelectPortal>
                        </Select>
                    }
                    </>
                )
            }) }

            <Button style={styles.margin}>
            <ButtonText>Submit</ButtonText>
            </Button>
        </FormControl>
    );
});

const styles = StyleSheet.create({
    margin: {
        margin: 5,
        marginLeft: 0,
    },
});

export default FormRender;