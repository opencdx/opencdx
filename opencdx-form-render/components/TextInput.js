import React from 'react';

import { View, TextInput as RNTextInput, TextInputProps as RNTextInputProps, Text, StyleSheet } from 'react-native';
import { useController, useFormContext } from 'react-hook-form';
const ControlledInput = (props) => {
    const formContext = useFormContext();
    const { formState } = formContext;
    const { name, label, defaultValue, ...inputProps } = props;
    const { field } = useController({ name, defaultValue });
    const hasError = Boolean(formState?.errors[name]);
    return (
        <View style={styles.container}>
            <View>
                <RNTextInput
                    autoCapitalize="none"
                    textAlign="left"
                    style={styles.input}
                    onChangeText={field.onChange}
                    onBlur={field.onBlur}
                    value={field.value}
                    {...inputProps}
                />
                <View style={styles.errorContainer}>
                    {hasError && (<Text style={styles.error}>{formState.errors[name].message}</Text>)}
                </View>
            </View>
        </View>
    );
}
export const TextInput = (props) => {
    const { name, rules, label, defaultValue, setFormError, ...inputProps } = props;
    const formContext = useFormContext();
    if (!formContext || !name) {
        const msg = !formContext ? "TextInput must be wrapped by the FormProvider" : "Name must be defined"
        setFormError(true)
        return null
    }
    return <ControlledInput {...props} />;
};
const styles = StyleSheet.create({
    label: {
        margin: 20,
        marginLeft: 0,
    },
    container: {
        flex: -1,
        justifyContent: 'center',
        padding: 8,
        borderColor: 'white',
        borderWidth: 1
    },
    input: {
        borderColor: 'gray',
        borderWidth: 1,
        padding: 10,
        borderRadius: 5,
    },
    errorContainer: {
        flex: -1,
        height: 25
    },
    error: {
        color: 'red'
    }
});