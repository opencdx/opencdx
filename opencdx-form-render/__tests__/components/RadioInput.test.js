import React, { useState } from 'react';
import renderer from 'react-test-renderer';
import { RadioInput } from '../../components/RadioInput';

describe('RadioInput', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const answerOption = [
    {
      id: 1,
      label: 'Yes',
      value: 'yes',
    },
    {
      id: 2,
      label: 'No',
      value: 'no',
    },
    {
      id: 3,
      label: 'Not Answered',
      value: 'not answered',
    }
  ];

  it('renders all answer options correctly', () => {
    const setFormErrorMock = jest.fn();

    const component = renderer.create(<RadioInput options={answerOption} setFormError={setFormErrorMock} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  
});

  