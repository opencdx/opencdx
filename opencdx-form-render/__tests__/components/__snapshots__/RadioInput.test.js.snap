// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioInput renders all answer options correctly 1`] = `null`;
